app/services/course/assessment/answer/text_response_comprehension_auto_grading_service.rb

```diff
@@ -0,0 +1,269 @@
+      end
+    end
+  end
+
+  # Find for all compre_keyword in answer.
+  # If word is found, set +answer_text_lemma_status["compre_keyword"][index]+ to the
+  # corresponding Solution.
+  # and collate an array of all Solutions where keywords are found in answer.
+  #
+  # @param [Array<String>] answer_text_lemma_array The lemmatised answer text in array form.
+  # @param [Hash{String=>Array<nil or TextResponseComprehensionPoint or TextResponseComprehensionSolution>}]
+  #   answer_text_lemma_status The status of each element in +answer_text_lemma+.
+  # @param [Hash{String=>Array<Course::Assessment::Question::TextResponseComprehensionSolution>}] hash
+  #   The mapping from keywords to Solutions.
+  def find_compre_keyword_in_answer(answer_text_lemma_array, answer_text_lemma_status, hash)
+    answer_text_lemma_array.each_index do |index|
+      next unless answer_text_lemma_status['compre_lifted_word'][index].nil?
+
+      # not a lifted word
+      answer_text_lemma_word = answer_text_lemma_array[index]
+      next unless hash.key?(answer_text_lemma_word) && !hash[answer_text_lemma_word].empty?
+
+      # keyword found in answer
+      until hash[answer_text_lemma_word].empty?
+        first_solution = hash[answer_text_lemma_word].shift
+        first_solution_point = first_solution.point
+
+        # for same Solution, remove from all other values in hash
+        hash.each_value do |solution_array|
+          solution_array.delete_if { |solution| solution.equal? first_solution }
+        end
+
+        unless answer_text_lemma_status['compre_lifted_word'].include?(first_solution_point)
+          # keyword (Solution) does NOT belong to a "lifted" Point
+          answer_text_lemma_status['compre_keyword'][index] = first_solution
+          break
+        end
+      end
+    end
+  end
+
+  # Returns the grade for a question with all matched solutions.
+  #
+  # The grade is considered to be the sum of grades assigned to all matched solutions, but not
+  # exceeding the maximum grade of the point, group and question.
+  #
+  # @param [Course::Assessment::Question::TextResponse] question The question answered by the
+  #   student.
+  # @param [Hash{String=>Array<nil or TextResponseComprehensionPoint or TextResponseComprehensionSolution>}]
+  #   answer_text_lemma_status The status of each element in +answer_text_lemma+.
+  # @return [Integer] The grade of the student answer for the question.
+  def grade_for(question, answer_text_lemma_status)
+    lifted_word_points = answer_text_lemma_status['compre_lifted_word']
+    keyword_solutions = answer_text_lemma_status['compre_keyword']
+
+    grade = 0
+    question.groups.each do |group|
+      group_grade = 0
+      group.points.each do |point|
+        next if lifted_word_points.include?(point)
+
+        solutions_found = point.solutions.select(&:compre_keyword?).map { |s| keyword_solutions.include?(s) }
+        (group_grade += point.point_grade) if solutions_found.all?
```

Comment:
Instead of computing `keyword_solutions.include?(s)` for all solutions, you can pass it directly to `all?` so that it can short-circuit on the first failing element.